contracts/tvl-staking/RibbonDualStakingRewards.sol

```diff
@@ -0,0 +1,355 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+// Inheritance
+import "../interfaces/IDualStakingRewards.sol";
+import "./RewardsDistributionRecipient.sol";
+import "../common/Pausable.sol";
+
+// https://docs.synthetix.io/contracts/source/contracts/stakingrewards
+contract DualStakingRewards is
+  IDualStakingRewards,
+  RewardsDistributionRecipient,
+  ReentrancyGuard,
+  Pausable
+{
+  using SafeMath for uint256;
+  using SafeCast for uint256;
+  using SafeERC20 for IERC20;
+
+  struct Rewards {
+    uint128 token0;
+    uint128 token1;
+  }
+
+  /* ========== STATE VARIABLES ========== */
+
+  IERC20 public rewardsToken0;
+  IERC20 public rewardsToken1;
+  IERC20 public stakingToken;
+  Rewards public rewardRate;
+  Rewards public rewardPerTokenStored;
+  uint256 public periodFinish = 0;
+  uint256 public rewardsDuration = 28 days;
+  uint256 public lastUpdateTime;
+
+  // timestamp dictating at what time of week to release rewards
+  // (ex: 1619226000 is Sat Apr 24 2021 01:00:00 GMT+0000 which will release as 1 am every saturday)
+  uint256 public startEmission;
+
+  mapping(address => Rewards) public userRewardPerTokenPaid;
+  mapping(address => Rewards) public rewards;
+
+  uint256 private _totalSupply;
+  mapping(address => uint256) private _balances;
+
+  /* ========== CONSTRUCTOR ========== */
+
+  constructor(
+    address _owner,
+    address _rewardsDistribution,
+    address _rewardsToken0,
+    address _rewardsToken1,
+    address _stakingToken,
+    uint256 _startEmission
+  ) Owned(_owner) {
+    require(_owner != address(0), "Owner must be non-zero address");
+    require(
+      _rewardsToken0 != address(0),
+      "Rewards token must be non-zero address"
+    );
+    require(
+      _rewardsToken1 != address(0),
+      "Rewards token must be non-zero address"
+    );
```

Comment:
More assertions:

```solidity
require(_rewardsToken0 != _stakingToken && _rewardsToken0 != _stakingToken);
require(_rewardsToken0 != _rewardsToken1);
```

Reply:
Good idea, added it here: https://github.com/ribbon-finance/governance/pull/46/commits/75b472db2a9fc817be19f04398c50a96b527837f